### ISSUE TITLE
How to start a graphical desktop environment inside the virtual environment of cubic that can be done forwarding X through ssh....

### ISSUE THREAD

**Issue by Marietto2008** [PJ-Singh-001/Cubic#95]
Hello.

I'm using Cubic on Ubuntu 22.04. It's good,I like it,but It misses an important feature. When I'm inside the virtual environment,I can't start a graphical desktop manager that will help a LOT to customize the iso. I know why I can't,but I don't know how to achieve this goal. Basically what's wrong is the network configuration. The virtual environment lacks of an internal IP number,that can be given to it using a different network connection scheme,such a network bridge. Instead at the moment inside of it I have the same network interfaces of the host OS ad when I try to forwarding X through ssh I will be connected to the host IP. Is there a way to correct this ? There are a lot of elements on the graphical interface of the OS that I want to customize that can't be easily configured ONLY through the Terminal. If I could do it moving them through a graphical desktop manager, it will help a LOT with the customizing. Im using Cubic 2022.06.72.

**Comment by datavectors** [comment]
> There are a lot of elements on the graphical interface of the OS that I want to customize that can't be easily configured ONLY through the Terminal.

What springs to my mind as a workaround is using an automation tool to modify UI elements. A simple (but dated) app such as [Actiona](https://github.com/Jmgr/actiona) in Ubuntu repo can help as a background parallel process. But I have not used it on Cubic UI. Or look at xdotool.

**Comment by Marietto2008** [comment]
what do you mean with "background parallel process" ? I would like to forward Xorg by installing ssh inside the cubic virtual environment and then I will enable to forwarding. In this way,I can configure easily the graphical elements of the cubic virtual environment directly inside of it. Actiona and xdotool can't work within the cubic virtual enviroment,because the desktop manager can't run inside of it.

**Comment by datavectors** [comment]
Perhaps I misunderstood but the basic idea was to "drive" the Cubic virtual environment from outside rather than within. Outside orchestration rather than inside. But just an idea which I have used on several other desktop apps to be orchestrated.

**Comment by Marietto2008** [comment]
I know what you mean.Your idea is to configure a desktop manager outside of the cubic vm and to copy its files inside of it. To be honest I don't like very much this method. I think that I can have a better control if I can orchestrate it from inside. And I don't need to configure an external OS if I can do it directly inside cubic. You know,my host os can be different than the os that I want to configure. And to run a qemu virtual machine to install the OS that I want to configure is a wasting of resources and time,if I can do everything inside cubic and it adds an extra layer of complexity. I suspect that the configuration process is faster orchestrating it from inside.

**Comment by datavectors** [comment]
Then another idea might be to embed an automation script (perhaps using Ansible playbooks) to further customise after Cubic spits out its ISO. That is, a two stage process.

**Comment by Marietto2008** [comment]
Do you realize that you are looking for more and unuseful complexity ? To create the better product,the author could put his hands inside the code and create an internal network stack with its IP number that can be used to run a desktop manager that can be configured inside the cubic environment. Are you the creator of cubic ? Or maybe another nice idea to get the same result,but not using the network stack. The full stop here is to think how to make the orchestration from within cubic.

**Comment by datavectors** [comment]
I am not part of Cubic team or its creator. Just a guy with ideas. passing by and read your question.
But why are you so closed to considering wildcard ideas?
You want to integrate some apps into a custom OS.
But there are more ways of skinning a cat.

The basic idea stems from a well developed principle of orchestrating microservices.
You can build on this principle.

You can regard Cubic workflow as just one microservice. But other microservices can work in concert (hence orchestration).

Here is one useful reference.

https://redis.com/docs/redis-microservices-for-dummies/

My suggestion was that you build an orchestrator. Cubic might be just one player in the orchestra.

Your closed thinking is “the author could put his hands inside the code”. A monolith.

https://answers.launchpad.net/cubic/+question/671192

And if you don't think your idea is a tad complex read here

https://unix.stackexchange.com/questions/12755/how-to-forward-x-over-ssh-to-run-graphics-applications-remotely

The idea about a two stage orchestration of Cubic is also raised here as a workaround.

https://askubuntu.com/questions/1057289/configuring-services-in-cubic-chroot

“Is it feasible to pre-configure a custom service to be running on first boot (or first login) after installing an .iso made with Cubic?”

Good luck.


After thought. You are pinning all your design on Cubic. How do you know if Cubic will always be available in the future? Recently I lost a lot of work by GitHub pulling the plug ("sunsetting") Atom editor  in favour of VSCode. Do not put all your eggs in one basket is the hard lesson I learned.

**Comment by Marietto2008** [comment]
I don't want to integrate some apps,I want to configure the menus,windows,decorations,taskbars positions of the wanted desktop manager in real time inside cubic. Why ? because I find it very comfortable to do while I'm inside cubic virtual environment. below you see that I tried to enable the X forwarding through ssh because I wanted to configure a desktop manager like lxde or xfce from within cubic,but it failed. Do you have a workaround ?

within cubic :

```
root@cubic:~# /etc/init.d/ssh start
 * Starting OpenBSD Secure Shell server sshd                                                  sshd: no hostkeys available -- exiting.
                                                                                       [fail]
root@cubic:~# ssh-keygen -A
ssh-keygen: generating new host keys: RSA DSA ECDSA ED25519 

root@cubic:~# /etc/init.d/ssh start
 * Starting OpenBSD Secure Shell server sshd                                           [ OK ] 

root@cubic:~# adduser test
Adding user `test' ...
Adding new group `test' (1000) ...
Adding new user `test' (1000) with group `test' ...
Creating home directory `/home/test' ...
Copying files from `/etc/skel' ...
New password: 
Retype new password: 
passwd: password updated successfully
Changing the user information for test
Enter the new value, or press ENTER for the default
 Full Name []: 
 Room Number []: 
 Work Phone []: 
 Home Phone []: 
 Other []: 
Is the information correct? [Y/n] y
```

on the host os :

```
ziomario@Z390-AORUS-PRO-DEST:~$ ssh -Y test@192.168.1.4 (host os IP Number)

The authenticity of host '192.168.1.4 (192.168.1.4)' can't be established.
This key is not known by any other names
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added '192.168.1.4' (ED25519) to the list of known hosts.
test@192.168.1.4's password: 
Permission denied, please try again.
```

it fails because the user test is not on the host os but on the cubic os,but **192.168.1.4** is the IP number of the 
host os (cubic has none). If I do : **ssh -Y test@192.168.1.4** inside cubic it will fails again because the user test is inside cubic,but the IP number is on the host os. If I create the user test on the host os and I will try to forward X using the host IP number and the user test,I will be able to configure xfce4 installed on the host os and that's not what I want.

**Comment by Marietto2008** [comment]
I've created this post to ask for some help :

https://unix.stackexchange.com/questions/715335/how-to-forward-x-over-ssh-to-configure-a-desktop-manager-within-cubic-remotely

**Comment by PJ-Singh-001** [comment]
This has been open for a while, and I haven’t seen additional input other than the good insights and recommendations from @datavectors. (Thanks @datavectors!)

@Marietto2008, regarding your desire to “configure the menus, windows, decorations, taskbars,” be aware that running a graphical interface to (attempt) to do this will not work the way you envision.

1. The GUI tools you typically use to configure themes, styles, etc. typically change the user’s settings. When you are on [Cubic’s Terminal page](https://github.com/PJ-Singh-001/Cubic/wiki/Terminal-Page), you are simply a root user who has read / write access to an isolated file system. I think you are confusing this for an actual running OS. Because the [terminal environment](https://github.com/PJ-Singh-001/Cubic/wiki/Terminal-Page) is not a running OS, services do not run, and you can not expect these back-end services to be available to help you configure your graphical settings using a front-end GUI tool. Even if you were able to forward to X and run a GUI tool, there is no back-end for your GUI tool to talk to!

2. Keep in mind users have not been created when you are customizing the OS in Cubic. So which user would you be changing the themes, settings, etc. for? If you want to do this globally, then you can not use GUI tools, you I will need to directly change the files that the OS uses to set the settings for _all_ users, or for the OS overall.

3. Most GUI tools leave behind files on your system when they are run. These may be configuration files that specify the size and location of a window, the last folder or directory that was accessed, or log files, etc. This may include SSH keys, IP address, user IDs, etc. If you are customizing a brand new OS, why would you want these superfluous files hanging around? They will just be copied onto _every_ machine your OS is installed onto.

Instead, I encourage you to think of Cubic as a way to _modify files_ in the Linux file system.

Place the correct information in the correct files, so that when the OS actually boots (either in a Live Environment or on an installed system), it behaves the way you want it to.

If you want a graphical tool to modify various files, use an editor on your host computer, such as Gedit, to create the necessary files, and then drag and drop them into the [Cubic terminal window](https://github.com/PJ-Singh-001/Cubic/wiki/Terminal-Page) to copy them into the current directory. (This way you do not have to use a command line editor, such as `nano`).

In summary, I understand what you _wish_ to do, but this will not work the way you want or expect it to, and you will be disappointed that your customizations were not reflected in your resulting ISO.

**Comment by Marietto2008** [comment]
Here I am. Please don't close my request,leave it open,I lke to talk about my ideas and to hear opinions,to share and learn. So,cubic as you have imagined it to be until now,allows the orchestrator to create all the users he/she wants,right ? but this task,on the configuration stage,is limited to only few steps. We can create a new user with the command "adduser newuser" and nothing more. What about,at this point,if cubic installs and run,automatically,some basic gui apps and services,like for sure,Xorg and wayland and some basic gui,like lxde under the root account ? The orchestrator can use these tools to continue configuring the system using a basic desktop manager,instead of doing the same things using the terminal window. The orchestrator will finish the configuration later,when the basic system will have been installed.
